### PR TITLE
'routing_engine' argument added to Device class which will allow to connec…

### PR DESCRIPTION
**Pull request by gaddevijay**
'routing_engine' argument added to Device class which will allow to connect to desired RE
Values to this argument can be master/backup/re0/re1/any

### CODE REVIEW

File: lib/jnpr/junos/facts/chassis.py [hunk 1]
```diff
@@ -41,6 +46,16 @@ def facts_chassis(junos, facts):
         facts['2RE'] = False
         x_ch = rsp.find('chassis')
 
+    if x_ch is None:  # This is when backup RE is being connected
+        try:
+            rsp = junos.rpc.get_software_information()
+            if rsp.tag == 'error':
+                raise RuntimeError()
```
Comment (vnitinv) [inline lib/jnpr/junos/facts/chassis.py]: Add message in runtime error. Or define an appropriate class exception for such cases.

File: lib/jnpr/junos/device.py [hunk 2]
```diff
@@ -441,6 +528,18 @@ def open(self, *vargs, **kvargs):
...
 
+            # If routing_engine(_RE) is defined then check if you are
+            # connected to correct RE, else get name of other re and
+            # connect to it.
+            self.connected = True
+
+            gather_facts = kvargs.get('gather_facts', self._gather_facts)
+            if gather_facts is True and self._RE is not None:
+                self._connected_re = _get_connected_slot()
+            if self._RE:
```
Comment (vnitinv) [inline lib/jnpr/junos/device.py]: check the logic

File: lib/jnpr/junos/device.py [hunk 3]
```diff
@@ -291,6 +293,12 @@ def __init__(self, *vargs, **kvargs):
         :param int port:
             *OPTIONAL* NETCONF port (defaults to 830)
 
+        :param str routing_engine:
+            *OPTIONAL* If not provided default behavior will be raise exception
```
Comment (vnitinv) [inline lib/jnpr/junos/device.py]: change the doc string a bit as discussed

File: lib/jnpr/junos/device.py [hunk 4]
```diff
@@ -412,6 +422,99 @@ def open(self, *vargs, **kvargs):
...
+        #  PRIVATE Functions
+        #
+
+        def _get_connected_slot():
+            """return connected slot. In case of error return None"""
+
+            rsp = self.rpc.get_chassis_inventory()
+            if rsp.tag == 'error':
+                raise RuntimeError('Could not determine RE slot id')
+            if rsp.tag == 'output':
```
Comment (stacywsmith) [inline lib/jnpr/junos/device.py]: I can confirm this is an incorrect assumption for an MX VC. On an MXVC, where each member has dual REs, all 4 routing engines successfully execute the get-chassis-inventory RPC.

File: lib/jnpr/junos/device.py [hunk 5]
```diff
@@ -412,6 +422,99 @@ def open(self, *vargs, **kvargs):
...
+                self._reRole = "backup"
+            else:
+                self._reRole = "master"
+
+            rsp = self.rpc.get_route_engine_information()
+            if rsp.tag == 'error':
+                raise RuntimeError('Could not determine RE slot id')
+            re_list = rsp.xpath('.//route-engine')
+            if len(re_list) == 1:
+                return "RE0"
```
Comment (stacywsmith) [inline lib/jnpr/junos/device.py]: This is an incorrect assumption. While uncommon, a chassis can have a single RE installed in slot 1. If there's only one RE, you should still look at the <slot> tag to determine the correct slot.

File: lib/jnpr/junos/device.py [hunk 6]
```diff
@@ -412,6 +422,99 @@ def open(self, *vargs, **kvargs):
...
+            for re in re_list:
```
Comment (stacywsmith) [inline lib/jnpr/junos/device.py]: This loop doesn't correctly handle any multi-chassis system.

File: lib/jnpr/junos/device.py [hunk 7]
```diff
@@ -412,6 +422,99 @@ def open(self, *vargs, **kvargs):
...
+
+            # Determine other re
+            other_re = None
+            if self._connected_re.upper() == 'RE0':
+                other_re = 're1'
+            else:
+                other_re = 're0'
+
+            # get the name of other re
+            command = etree.XML('<configuration><groups><name>%s</name>'
```
Comment (stacywsmith) [inline lib/jnpr/junos/device.py]: This assumes the RE has a unique hostname configured, and the host executing the PyEZ script can resolve this unqualified hostname to a unique IP address, has connectivity to that IP address, etc. There are a LOT of assumptions here which I don't think are safe to make.

**Comment by jnpr-community-netdev** [comment]
Autobot:  Would an admin like to run functional tests?

**Comment by vnitinv** [comment]
Can one of the admins verify this patch?

**Comment by vnitinv** [comment]
Thanks for the pull request. will do it properly in another pull request in future.

**Pull request closed without merging**